libbeat/reader/multiline/counter.go

```diff
@@ -0,0 +1,133 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package multiline
+
+import (
+	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/elastic/beats/v7/libbeat/reader"
+)
+
+type counterReader struct {
+	reader    reader.Reader
+	state     func(*counterReader) (reader.Message, error)
+	logger    *logp.Logger
+	msgBuffer *messageBuffer
+}
+
+func newMultilineCountReader(
+	r reader.Reader,
+	separator string,
+	maxBytes int,
+	config *Config,
+) (reader.Reader, error) {
+	return &counterReader{
+		reader:    r,
+		state:     (*counterReader).readFirstCount,
+		msgBuffer: newMessageBuffer(maxBytes, config.LinesCount, []byte(separator), config.SkipNewLine),
+		logger:    logp.NewLogger("reader_counter_multiline"),
+	}, nil
+}
+
+// Next returns next multi-line event.
+func (cr *counterReader) Next() (reader.Message, error) {
+	return cr.state(cr)
+}
+
+func (cr *counterReader) readFailedCount() (reader.Message, error) {
+	err := cr.msgBuffer.err
+	cr.msgBuffer.setErr(nil)
+	cr.resetCountState()
+	return reader.Message{}, err
+}
+
+// resetCountState sets state of the reader to readFirst
+func (cr *counterReader) resetCountState() {
+	cr.setState((*counterReader).readFirstCount)
+}
+
+func (cr *counterReader) readFirstCount() (reader.Message, error) {
```

Comment:
I wonder if we can simplify readNextCount and make it more eager (the pattern based multiline matcher postpones some operations to the next line because it has not enough information yet). With `readNextCount` getting more simple we might not need `readFirstCount` and maybe even not `readFailedCount`.